eth/handler.go

```diff
@@ -250,8 +250,17 @@ func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
 		peer.Log().Debug("Ethereum handshake failed", "err", err)
 		return err
 	}
+	reserved := 0 // reserved peer slots
+	if atomic.LoadUint32(&h.snapSync) == 1 && !peer.SupportsCap("snap", 1) {
+		// If we are running snap-sync, we want to reserve half the peer slots for
+		// peers supporting the snap protocol
+		reserved = h.maxPeers/2 - len(h.peers.snapPeers)
+	}
+	if reserved < 0 {
+		reserved = 0
+	}
```

Comment:
I like it that the solution is short and sweet, but there's a cornercase we need to support. Currently the P2P layer splits the allowed peer count into 2, reserving half for inbound and half for outbound connections. If my node is NATed, I won't get any inbound, so I'll use up 1/2 of my slots outbound. The issue is that I'm able to saturate my outbound connections fully with `eth` only peers without triggering the above snap reservation.

One solution is that we reserve half of the inbound and half of the outbound connections separately for snap. That would solve the above mentioned issue, but it can still be a bit flaky if I only have 1/4 peer.

Another alternative could be to enforce the restriction based on the currently connection peer count. E.g. `if peers.LenEth() > min(2 * peers.LenSnap(), 5)`, then reject. That would guarantee that we can always get some baseline number of eth peers, but above that, we're enforicng half and half.

Reply:
Not quite sure I understand... Do you mean like this?
```
		snapPeers := h.peers.SnapLen()
		ethPeers := h.peers.Len() - snapPeers // snap-peers are a subset of eth-peers
		if ethPeers > 5 && ethPeers > 2 *snapPeers{
			reject = true
		}
```
In that case, it would always accept up to `5` non-snap peers, but would otherwise allow twice as many non-snap as snap. 

Alternatively; 
```
		snapPeers := h.peers.SnapLen()
		ethPeers := h.peers.Len() - snapPeers // snap-peers are a subset of eth-peers
		if ethPeers > 5 && ethPeers > snapPeers{
			reject = true
		}
```

In that case, it would always accept up to `5` non-snap peers, but would otherwise allow only as many non-snap as snap.

Reply:
I guess a different method would be to "reject if non-snap peers are more than 5+snap-peers". 

As it is currently (`83a9567`), going from `5` to you need `3` snap peers to join. before you allow the `6th` non-snap peer. 
The alternative would only require `1` snap-peer before the `6th` is allowed in. And eventually, with `50` peers, you'll have something like `23` snap and `27` non-snap peers.